File: app/src/main/java/com/example/all_sensors/presentation/AccListener.java
package com.example.all_sensors.presentation;

import android.util.Log;

import com.example.all_sensors.presentation.data.DataBuffer;
import com.samsung.android.service.health.tracking.HealthTracker;
import com.samsung.android.service.health.tracking.data.DataPoint;
import com.samsung.android.service.health.tracking.data.ValueKey;

import java.util.List;
import java.util.Locale;

/**
 * Listener para el acelerómetro continuo.
 * Aquí puedes enviar los datos a tu DataBuffer u Observer.
 */
public class AccListener implements HealthTracker.TrackerEventListener {
    private final DataBuffer buffer;

    public AccListener(DataBuffer buffer) {
        this.buffer = buffer;
    }
    @Override
    public void onDataReceived(List<DataPoint> dataPoints) {
        long now = System.currentTimeMillis();
        // dataPoints contendrá AccelerometerSet con X,Y,Z a 25Hz extraes cada datapoint con samsung
        // y conviertes a data point
        for (com.samsung.android.service.health.tracking.data.DataPoint dp : dataPoints) {
            float x = dp.getValue(ValueKey.AccelerometerSet.ACCELEROMETER_X);
            float y = dp.getValue(ValueKey.AccelerometerSet.ACCELEROMETER_Y);
            float z = dp.getValue(ValueKey.AccelerometerSet.ACCELEROMETER_Z);

            Log.d("ACC_DATA",
                    String.format(Locale.US, "%d — x:%.2f, y:%.2f, z:%.2f", now, x, y, z));

            String vals = String.format(Locale.US, "%.2f,%.2f,%.2f", x, y, z);
            buffer.add(new com.example.all_sensors.presentation.data.DataPoint(now, "ACC", vals));

            ObserverUpdater.getObserverUpdater().notifyAccObservers(x, y, z);
        }
    }

    @Override
    public void onFlushCompleted() {
        // No suele usarse para continuos
    }

    @Override
    public void onError(com.samsung.android.service.health.tracking.HealthTracker.TrackerError error) {
        // TODO: manejar error (por ejemplo permisos denegados)
        Log.e("ACC_DATA", "ACC error");
    }


}

File: app/src/main/java/com/signals/smartwatch/sensors/samsung/AccListener.kt
package com.signals.smartwatch.sensors.samsung

import android.util.Log
import com.signals.smartwatch.SensorEvents
import com.signals.smartwatch.data.SensorDao
import com.signals.smartwatch.data.SensorEntity
import com.samsung.android.service.health.tracking.HealthTracker
import com.samsung.android.service.health.tracking.data.DataPoint
import com.samsung.android.service.health.tracking.data.ValueKey

class AccListener(
    private val dao: SensorDao,
    private val events: SensorEvents
) : HealthTracker.TrackerEventListener {

    private var logCounter = 0

    override fun onDataReceived(dataPoints: List<DataPoint>) {
        val now = System.currentTimeMillis()
        dataPoints.forEach { dp ->
            val x = dp.getValue(ValueKey.AccelerometerSet.ACCELEROMETER_X)
            val y = dp.getValue(ValueKey.AccelerometerSet.ACCELEROMETER_Y)
            val z = dp.getValue(ValueKey.AccelerometerSet.ACCELEROMETER_Z)

            if (logCounter++ % LOG_EVERY_N == 0) {
                Log.d(TAG, "$now — x:${"%.2f".format(x)}, y:${"%.2f".format(y)}, z:${"%.2f".format(z)}")
            }

            dao.insertBlocking(SensorEntity(
                timestamp = now,
                sensor    = "ACC",
                values    = "%.2f,%.2f,%.2f".format(x, y, z)
            ))
            events.onAcc(x.toFloat(), y.toFloat(), z.toFloat())
        }
    }

    override fun onFlushCompleted() = Unit

    override fun onError(error: HealthTracker.TrackerError) {
        Log.e(TAG, "Error: $error")
    }

    companion object {
        private const val TAG = "ACC_DATA"
        private const val LOG_EVERY_N = 25
    }
}
